graphql_compiler/query_pagination/pagination_planning.py

```diff
@@ -1,53 +1,156 @@
 # Copyright 2019-present Kensho Technologies, LLC.
-from collections import namedtuple
+from typing import List, NamedTuple, Tuple
+
+from graphql import DocumentNode
 
 from ..ast_manipulation import get_only_query_definition, get_only_selection_from_ast
-from ..cost_estimation.helpers import is_int_field_type, is_uuid4_type
+from ..cost_estimation.helpers import is_uuid4_type
+from ..cost_estimation.int_value_conversion import field_supports_range_reasoning
 from ..exceptions import GraphQLError
+from ..schema.schema_info import QueryPlanningSchemaInfo
+
+
+class PaginationWarning(object):
+    def __init__(self, message):
+        """Initialize a PaginationWarning with a custom message."""
+        self.message = message
+
+
+class PaginationFieldNotSpecified(PaginationWarning):
+    def __init__(self, vertex_name):
+        """Initialize a PaginationFieldNotSpecified PaginationWarning."""
+        super(PaginationFieldNotSpecified, self).__init__(
+            "Pagination field not specified for vertex {}".format(vertex_name)
+        )
+        self.vertex_name = vertex_name
+
+
+class NotEnoughQuantiles(PaginationWarning):
+    def __init__(self, vertex_name, field_name, current_resolution, desired_resolution):
+        """Initialize a NotEnoughQuantiles PaginationWarning."""
+        super(NotEnoughQuantiles, self).__init__(
+            "Pagination would have been more successful if more quantiles were provided "
+            "for {}.{}. Currently there are {}, ideally there should be {}".format(
+                vertex_name, field_name, current_resolution, desired_resolution
+            )
+        )
+        self.vertex_name = vertex_name
+        self.field_name = field_name
+        self.current_resolution = current_resolution
+        self.desired_resolution = desired_resolution
+
+
+class VertexPartition(NamedTuple):
+    """Plan to split the query at a certain vertex into a certain number of pages."""
 
+    # field names leading to the vertex to be split
+    query_path: Tuple[str, ...]
 
-# The intent to split the query at a certain vertex into a certain number of pages.
-VertexPartition = namedtuple(
-    "VertexPartition",
-    (
-        "query_path",  # Tuple[field name : str] leading to the vertex to be split
-        "number_of_splits",  # The number of subdivisions intended for this vertex
-    ),
-)
+    # field to use for pagination
+    pagination_field: str
 
+    # The number of subdivisions intended for this vertex
+    number_of_splits: int
 
-# The intent to split the query with a combination of VertexPartitions
-PaginationPlan = namedtuple("PaginationPlan", ("vertex_partitions",))  # List[VertexPartition]
 
+class PaginationPlan(NamedTuple):
```

Comment:
Since `get_pagination_plan` might not always be able to paginate to the desired number of pages, it would be nice if the pagination plan specified the number of pages it contains.

Reply:
To avoid denormalization problems, I added a function that computes that instead.